### PR TITLE
Proto nn dev

### PR DESCRIPTION
**Pull request by harsha-simhadri**

### CODE REVIEW

**Comment by sureshiyengar** [comment]
Looks good to me.

File: ProtoNNLocalPredictor.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  assert(sizeof(MKL_INT) == 8 && "need large enough indices to store matrices");
+  assert(sizeof(MKL_INT) == sizeof(Eigen::Index) && "MKL BLAS routines are called directly on data of an Eigen matrix. Hence, the index sizes should match."); 
+
+  EdgeML::ProtoNN::ProtoNNPredictor predictor(EdgeML::DataIngestType::FileIngest,
+                                              argc, (const char**)argv);
+
+  ProtoNNPredictor::ResultStruct res;
+  res = predictor.evaluate();
+  switch(res.problemType) {
+    case binary:
```
Comment (lodenrogue) [inline ProtoNNLocalPredictor.cpp]: Do we want this empty case here?

File: src/ProtoNN/ProtoNNPredictor.cpp [hunk 2]
```diff
@@ -38,9 +140,126 @@ ProtoNNPredictor::ProtoNNPredictor(
 #endif
 }
 
+void ProtoNNPredictor::createOutputDirs()
```
Comment (harsha-simhadri) [inline src/ProtoNN/ProtoNNPredictor.cpp]: @t-chgupt @adityakusupati Can we agree on a common output dir name for ProtoNN and Bonsai and move this code to common

File: src/ProtoNN/ProtoNNPredictor.cpp [hunk 3]
```diff
@@ -38,9 +140,126 @@ ProtoNNPredictor::ProtoNNPredictor(
...
+#endif
+#endif
+  }
+  catch (...) {
+    LOG_WARNING("Error in creating one of the subdirectories. Some of the output may not be recorded.");
+  }
+}
+
+
+void ProtoNNPredictor::setFromArgs(const int argc, const char** argv)
```
Comment (harsha-simhadri) [inline src/ProtoNN/ProtoNNPredictor.cpp]: @t-chgupt  If the user does not provide sufficient arguments, do we assert and fail

File: src/ProtoNN/ProtoNNPredictor.cpp [hunk 4]
```diff
@@ -155,7 +376,273 @@ void ProtoNNPredictor::scoreSparseDataPoint(
...
+  
+  return res;
+}
+
+
+// computes accuracy for binary/multiclass datasets, and prec1, prec3, prec5 for multilabel datasets
+ProtoNNPredictor::ResultStruct ProtoNNPredictor::evaluateBatch(
+  const MatrixXuf& Yscores, 
+  const LabelMatType& Y)
+{
```
Comment (harsha-simhadri) [inline src/ProtoNN/ProtoNNPredictor.cpp]: @t-chgupt @adityakusupati can we move it to common and reuse?

File: src/ProtoNN/ProtoNNPredictor.cpp [hunk 5]
```diff
@@ -9,6 +9,108 @@
 using namespace EdgeML;
 using namespace EdgeML::ProtoNN;
 
+
```
Comment (harsha-simhadri) [inline src/ProtoNN/ProtoNNPredictor.cpp]: @t-chgupt  Can we move constructor to the top

File: src/ProtoNN/ProtoNNPredictor.cpp [hunk 6]
```diff
@@ -9,6 +9,108 @@
...
+  precision3 *= scale;
+  precision5 *= scale;
+}
+
+ProtoNNPredictor::ProtoNNPredictor(
+  const DataIngestType& dataIngestType,
+  const int& argc,
+  const char ** argv)
+{
+  assert(dataIngestType == FileIngest);
```
Comment (harsha-simhadri) [inline src/ProtoNN/ProtoNNPredictor.cpp]: @t-chgupt  This is strange. If we require fileIngest, why ask for dataIngestType. Let's just document this constructor as being useful only for when model is read from file.

**Pull request merged**